plugins/outputs/zabbix/utils.go

```diff
@@ -0,0 +1,113 @@
+	errMetricNoTags            = fmt.Errorf("metric without tags")
+	errMetricJustHostTag       = fmt.Errorf("metric with only host tag")
+)
+
+// getHostname returns the hostname from the tags, or the system hostname if not found.
+func getHostname(hostTag string, metric telegraf.Metric) (string, error) {
+	if hostname, ok := metric.GetTag(hostTag); ok {
+		return hostname, nil
+	}
+
+	return os.Hostname()
+}
+
+// processMetric converts a Telegraf metric to a list of Zabbix metrics.
+// Ignore metrics with no hostname.
+func (z Zabbix) processMetric(metric telegraf.Metric) []*zabbix.Metric {
+	zbxMetrics := make([]*zabbix.Metric, 0, len(metric.FieldList()))
+
+	for _, field := range metric.FieldList() {
+		zbxMetric, err := z.buildZabbixMetric(metric, field.Key, field.Value)
+		if err != nil {
+			z.Log.Errorf("Error converting telegraf metric to Zabbix format: %v", err)
+			continue
+		}
+
+		zbxMetrics = append(zbxMetrics, zbxMetric)
+	}
+
+	return zbxMetrics
+}
+
+// buildZabbixMetric builds a Zabbix metric from a Telegraf metric, for one particular value.
+func (z Zabbix) buildZabbixMetric(metric telegraf.Metric, fieldName string, value interface{}) (*zabbix.Metric, error) {
+	hostname, err := getHostname(z.HostTag, metric)
+	if err != nil {
+		return nil, fmt.Errorf("error getting hostname: %w", err)
+	}
+
+	metricValue, err := internal.ToString(value)
+	if err != nil {
+		return nil, fmt.Errorf("error converting value: %w", err)
+	}
+
+	key := z.Prefix + metric.Name() + "." + fieldName
+	if z.SkipMeasurementPrefix {
+		key = z.Prefix + fieldName
+	}
+
+	// Ignore host tag.
+	// We want to add tags to the key in alphabetical order. Eg.:
+	// lld.dns_query.query_time_ms[DOMAIN,RECORD_TYPE,SERVER]
+	// TagList already return the tags in alphabetical order.
+	tagValues := make([]string, 0, len(metric.TagList()))
+
+	for _, tag := range metric.TagList() {
+		if tag.Key == z.HostTag {
+			continue
+		}
+
+		// Get tag values in the same order as the tag keys in the tags slice.
+		tagValues = append(tagValues, tag.Value)
+	}
+
+	if len(tagValues) != 0 {
+		key = fmt.Sprintf("%v[%v]", key, strings.Join(tagValues, ","))
+	}
+
+	return zabbix.NewMetric(hostname, key, metricValue, z.AgentActive, metric.Time().Unix()), nil
+}
+
+// lldHash returns an uniq identifier based on measurement and tags keys.
+// Format generated: Name.Tags_sorted.
+// Empty tags are ignored.
+func generateLLDKey(in telegraf.Metric, hostTag string) (string, error) {
+	name := in.Name()
+	tags := in.TagList()
+
+	if name == "" {
+		return "", errMetricNoMeasurementName
+	}
+
+	if len(tags) == 0 {
+		return "", errMetricNoTags
+	}
+
+	tagKeys := make([]string, 0, len(tags))
+	for _, tag := range tags {
+		if tag.Value != "" && tag.Key != hostTag {
+			tagKeys = append(tagKeys, tag.Key)
+		}
+	}
+
+	if len(tagKeys) == 0 {
+		return "", errMetricJustHostTag
+	}
+
+	return name + "." + strings.Join(tagKeys, "."), nil
+}
```

Comment:
I think this function is only used in the test. So please move the function there to exclude it from the Telegraf binary.

Reply:
fixed in 018b4c432